# HBIGS course 2021/22: Introduction to computational RNA biology
## Splicing, translation and networks

### Dates and venue

24.03.2022 from 9.00-17.00. Seminar room AZIII 03.434 (Analysezentrum 3, Im Neuenheimer Feld 669).

25.03.2022 from 9.00-17.00. Seminar room AZIII 03.434 (Analysezentrum 3, Im Neuenheimer Feld 669). 

> In addition, there will be a walk-in session to help you set up your laptop on 23.03.2022 from 16.00. 
> **Please note that it is your responsibility to make sure that you are ready for the course.** 
> If you have difficulties to install the required software or to access our HPC cluster, you must come to the walk-in session! 

### Description

This repository contains material for the HBIGS course 2022. The workshop's main objective is to learn how to use 
selected bioinformatics workflows and tools to analyse RNA-, Ribo-seq, and proteomics data. To attain this objective, participants will
be led through a series of practical exercises. 

We will be using a subset of the data from [D.A. Rothenberg et al. A Proteomics Approach to Profiling the Temporal Translational Response to Stress and Growth. iScience. 2018; 9:367-381](https://www.sciencedirect.com/science/article/pii/S2589004218301949?via%3Dihub). Alternatively, you can access the article [here](https://data.dieterichlab.org/s/PqSMgTLHJcJzstD).

### Requirements

#### Hardware/software requirements

You need a functional laptop with *e.g.* eduroam connection, and WireGuard installed (see [Setting Up WireGuard](https://github.com/dieterich-lab/hbigs_course_2022/wiki/Preparation-Workshop)). 

> If you have difficulties to install the required software or to access our HPC cluster, you must 
> come to the walk-in session on 23.03.2022.

#### General requirements

Before the course, please consult these resources if needed:

* [BASH shell basics](https://swcarpentry.github.io/shell-novice) or [command line tutorials](https://tutorials.ubuntu.com/tutorial/command-line-for-beginners)
* [Python basics](https://swcarpentry.github.io/python-novice-inflammation) - Sections 1 to 5
* [R Basics](http://swcarpentry.github.io/r-novice-inflammation) - Sections 1 to 4
* Some Linux editor like [nano](https://staffwww.fullcoll.edu/sedwards/Nano/IntroToNano.html) or [vim](https://www.openvim.com)

See also [further reading](https://github.com/dieterich-lab/hbigs_course_2022/wiki/Further-Reading).

**Note:** These links are being provided as a convenience and for informational purposes only; they do not constitute an endorsement or an approval by us of 
their content. We bear no responsibility for the accuracy of their content.

### Schedule

| Walk-in session<br>23.03.2022 | Topic | Instructor |
:----------|:-----------------|:----------|
| 16.00           | [Setting Up WireGuard](https://github.com/dieterich-lab/hbigs_course_2022/wiki/Preparation-Workshop)<br>[installing missing packages 1](part1_RNAseq/README.md#dependencies)<br>[installing missing packages 2](part4_Integration/README.md#dependencies)  | HW, EG, EB  |

| Day 1<br>24.03.2022 | Topic | Instructor |
:----------|:-----------------|:----------|
| 9.00-12.30 | [From raw sequencing reads to differential expression](part1_RNAseq/README.md) | EG, EB |
| 12.30-13.30 | Lunch |  |
| 13.30-15.00 | [From raw sequencing reads to differential expression](part1_RNAseq/README.md) | EG, EB |
| 15.00-17.00 | Approaches for differential splicing analysis from RNA-Seq data | TBB |

| Day 2<br>25.03.2022 | Topic | Instructor |
:----------|:-----------------|:----------|
| 9.00-12.30 | [Ribo-seq data analysis and the Rp-Bp workflow](part3_Riboseq/README.md) | EB |
| 12.30-13.30 | Lunch |  |
| 13.30-17.00 | [Proteomics and Integrative Analysis](part4_Integration/README.md) | EG |

### Contact

HW: Harald Wilhelmi (system administrator, <it@dieterichlab.org>)<br>
EG: Enio Gjerga (<e.gjerga@uni-heidelberg.de>)<br> 
EB: Etienne Boileau (<boileau@uni-heidelberg.de>)<br>
TBB: Thiago Britto Borges (<Thiago.BrittoBorges@uni-heidelberg.de>)



