---
title: "Analysis and Exploration of Proteomics Data"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

## Proteomics Data Analysis
Below are shown the steps typically used to perform Differential Gene Expression analysis. The analysis steps have been applied over the _EGF-driven protein synthesis_ transcriptomics data from [D.A. Rothenberg et al. A Proteomics Approach to Profiling the Temporal Translational Response to Stress and Growth. iScience. 2018; 9:367-381](https://www.sciencedirect.com/science/article/pii/S2589004218301949?via%3Dihub).

### Loading of R-Packages
We start by loading the R-packages we need to use for our analysis:

```{r chunk1, echo=TRUE, eval=TRUE, include=TRUE, warning=FALSE, comment=FALSE, message=FALSE}
library(edgeR)
library(lattice)
library(biomaRt)
library(gplots)
library(gridExtra)
library(openxlsx)
library(tools)
library(ggplot2)
library(dplyr)
library(Glimma)
library(openxlsx)
```

### Loading of the data
We load the generated **subread_counts** object, define the experimental conditions for each available sample and prepare the data in the _DGEObj_ format which represents a flexible container to manage and annotate Differential Gene Expression (DGE) analysis results.

```{r chunk2, echo=TRUE, eval=TRUE}

load("course.Rdata")

groups <- factor(c("PBS","EGF","PBS","EGF"))

label <- c("PBS2", "EGF1", "PBS1", "EGF2")

print(colnames(subread_counts$counts))
print(colnames(subread_counts$counts))

DGEObj <- DGEList(group=groups, counts=subread_counts$counts, genes=subread_counts$annotation[,c("GeneID","Length")])

print(DGEObj)
```

### Experimental Design
We make an experimental design matrix where we indicate the condition and replicate ID's for each sample. We then define the testing contrasts of our experimental design (*EGF_vs_PBS*). Our targeted *EGF_vs_PBS* contrast tests whether the average expression across all *EGF* groups is equal to the average expression across all *PBS* groups using _makeContrasts_.

```{r chunk3, echo=TRUE, eval=TRUE}
design <- model.matrix(~0+groups)
colnames(design) <- levels(groups)
print(design)

contrasts <- makeContrasts(
  EGF_vs_PBS = EGF - PBS,
  levels=design
)
counts <- DGEObj$counts
```

### DGEObj Processing
We simplify a bit the names of the samples.

```{r chunk4, echo=TRUE, eval=TRUE}
head(DGEObj$samples)

rownames(DGEObj$samples) <- gsub("\\.", "_", rownames(DGEObj$samples))
rownames(DGEObj$samples) <- gsub("KD_", "", rownames(DGEObj$samples))
rownames(DGEObj$samples) <- gsub("_circSLC8A1", "", rownames(DGEObj$samples))
rownames(DGEObj$samples) <- gsub("_bam", "", rownames(DGEObj$samples))
rownames(DGEObj$counts) <- gsub("_1_mRNA_Seq_PBS", "", rownames(DGEObj$counts))
rownames(DGEObj$counts) <- gsub("_1_mRNA_Seq_EGF", "", rownames(DGEObj$counts))
rownames(DGEObj$counts) <- gsub("_2_mRNA_Seq_PBS", "", rownames(DGEObj$counts))
rownames(DGEObj$counts) <- gsub("_2_mRNA_Seq_EGF", "", rownames(DGEObj$counts))

colnames(DGEObj$counts) <- gsub("\\.", "_", colnames(DGEObj$counts))
colnames(DGEObj$counts) <- gsub("KD_", "", colnames(DGEObj$counts))
colnames(DGEObj$counts) <- gsub("_1_mRNA_Seq_PBS", "", colnames(DGEObj$counts))
colnames(DGEObj$counts) <- gsub("_1_mRNA_Seq_EGF", "", colnames(DGEObj$counts))
colnames(DGEObj$counts) <- gsub("_2_mRNA_Seq_PBS", "", colnames(DGEObj$counts))
colnames(DGEObj$counts) <- gsub("_2_mRNA_Seq_EGF", "", colnames(DGEObj$counts))
colnames(DGEObj$counts) <- gsub("_bam", "", colnames(DGEObj$counts))


DGEObj$samples
head(DGEObj$counts)
head(DGEObj$genes)
DGEObj
```

### Data Filtering
We filter the data by keeping only candidates with > 2 counts per million mapped reads AND in 3 libs.

```{r chunk5, echo=TRUE, eval=TRUE}
keep <- rowSums(cpm(DGEObj) > 1) >= 3

# select sub set from above
DGEObj <- DGEObj[keep, keep.lib.sizes = FALSE]

dim(DGEObj)
```

### edgeR Analysis
We calculate the normalization factor, estimate the dispersion, fit the linear model, and then summarize the results in an edgeR table format.

```{r chunk6, echo=TRUE, eval=TRUE}
# The filtered raw counts are then normalized with calcNormFactors according to 
# the weighted trimmed mean of M-values (TMM) to eliminate composition biases 
# between libraries.
DGEObj <- calcNormFactors(DGEObj) # recalc norm factors - TMM

# With the normalized gene counts and design matrix we can now generate the 
# negative binomial (NB) dispersion estimates using the estimateDisp function. 
# The NB dispersion estimates reflect the overall biological variability under 
# the Quasi-Likelihood framework in edgeR.
DGEObj <- estimateDisp(DGEObj, design) # estimate dispersion
DGEObj$common.dispersion

# A generalized linear model (GLM) of the Negative Binomial family is used to 
# fit the data. 
fit <- glmFit(DGEObj, design) # fit generalized linear model
lrt <- glmLRT(fit, contrast = contrasts[, "EGF_vs_PBS"])

# main data table of edgeR results
edgerTable <- topTags(lrt, n = nrow(DGEObj))$table

```

### Mapping
We do mappping of the Ensembl gene ID's.

```{r chunk7, echo=TRUE, eval=TRUE}
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl", host="ensembl.org")

genemap <- getBM(attributes = c("external_gene_name", "ensembl_gene_id", "description"), 
                 values = edgerTable$GeneID, mart = ensembl)
genemap2 <- getBM(attributes = c("ensembl_gene_id", "go_id"), filter = "ensembl_gene_id", 
                  values = edgerTable$GeneID, mart = ensembl)

edgerTable.idx <- match(edgerTable$GeneID, genemap$ensembl_gene_id)
edgerTable.idx2 <- match(edgerTable$GeneID, genemap2$ensembl_gene_id)

# add biomart data to results tablesTRUE
edgerTable$external_gene_name <- genemap$external_gene_name[edgerTable.idx]
# edgerTable$entrezgene <- genemap$entrezgene[edgerTable.idx]
edgerTable$description <- genemap$description[edgerTable.idx]
edgerTable$GO <- genemap2$go_id[edgerTable.idx2]
edgerTable$ensembl_gene_id <- genemap$ensembl_gene_id[edgerTable.idx]
lrt$table$gene <- genemap$external_gene_name[edgerTable.idx]
gene_names <- data.frame(lrt$table$gene)

edgerTable <- edgerTable[c("external_gene_name", "GeneID" , "Length" , "logFC", 
                           "logCPM", "PValue", "FDR", "GO", "description")]

ttop_dge <- edgerTable
head(ttop_dge)
```